File: scripts/download-whisper-model.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');
const https = require('https');

// Create models directory
const modelsDir = path.join(__dirname, '..', 'public', 'models', 'whisper-tiny-en');
if (!fs.existsSync(modelsDir)) {
  fs.mkdirSync(modelsDir, { recursive: true });
}

// Whisper tiny.en model files from Hugging Face
const modelFiles = [
  'config.json',
  'generation_config.json',
  'merges.txt',
  'normalizer.json',
  'preprocessor_config.json',
  'tokenizer.json',
  'tokenizer_config.json',
  'vocab.json',
  'onnx/decoder_model_merged.onnx',
  'onnx/encoder_model.onnx'
];

const baseUrl = 'https://huggingface.co/Xenova/whisper-tiny.en/resolve/main/';

function downloadFile(url, filePath) {
  return new Promise((resolve, reject) => {
    const dir = path.dirname(filePath);
    if (!fs.existsSync(dir)) {
      fs.mkdirSync(dir, { recursive: true });
    }

    const file = fs.createWriteStream(filePath);
    
    console.log(`📥 Downloading: ${path.basename(filePath)}`);
    
    https.get(url, (response) => {
      if (response.statusCode === 200) {
        response.pipe(file);
        file.on('finish', () => {
          file.close();
          console.log(`✅ Downloaded: ${path.basename(filePath)}`);
          resolve();
        });
      } else if (response.statusCode === 302 || response.statusCode === 301) {
        // Handle redirects
        downloadFile(response.headers.location, filePath).then(resolve).catch(reject);
      } else {
        reject(new Error(`Failed to download ${url}: ${response.statusCode}`));
      }
    }).on('error', (err) => {
      fs.unlink(filePath, () => {}); // Delete the file on error
      reject(err);
    });
  });
}

async function downloadAllFiles() {
  console.log('🚀 Starting Whisper model download...');
  
  for (const file of modelFiles) {
    const url = baseUrl + file;
    const filePath = path.join(modelsDir, file);
    
    try {
      await downloadFile(url, filePath);
    } catch (error) {
      console.error(`❌ Failed to download ${file}:`, error.message);
      process.exit(1);
    }
  }
  
  console.log('🎉 All model files downloaded successfully!');
  console.log(`📁 Model location: ${modelsDir}`);
}

downloadAllFiles().catch(console.error);

File: scripts/download-whisper-model.py
#!/usr/bin/env python3

import os
import requests
import json
from pathlib import Path

def download_file(url, filepath):
    """Download a file from URL to filepath"""
    print(f"📥 Downloading: {filepath.name}")
    
    response = requests.get(url, stream=True)
    response.raise_for_status()
    
    # Create directory if it doesn't exist
    filepath.parent.mkdir(parents=True, exist_ok=True)
    
    with open(filepath, 'wb') as f:
        for chunk in response.iter_content(chunk_size=8192):
            f.write(chunk)
    
    print(f"✅ Downloaded: {filepath.name}")

def main():
    # Create models directory
    models_dir = Path(__file__).parent.parent / "public" / "models" / "whisper-tiny-en"
    models_dir.mkdir(parents=True, exist_ok=True)
    
    # Whisper tiny.en model files from Hugging Face
    model_files = [
        'config.json',
        'generation_config.json',
        'merges.txt',
        'normalizer.json',
        'preprocessor_config.json',
        'tokenizer.json',
        'tokenizer_config.json',
        'vocab.json',
        'onnx/decoder_model_merged.onnx',
        'onnx/encoder_model.onnx'
    ]
    
    base_url = 'https://huggingface.co/Xenova/whisper-tiny.en/resolve/main/'
    
    print('🚀 Starting Whisper model download...')
    
    for file in model_files:
        url = base_url + file
        filepath = models_dir / file
        
        try:
            download_file(url, filepath)
        except Exception as e:
            print(f"❌ Failed to download {file}: {e}")
            return 1
    
    print('🎉 All model files downloaded successfully!')
    print(f'📁 Model location: {models_dir}')
    
    # Create a simple index file to verify the model is available
    index_data = {
        "model_name": "whisper-tiny.en",
        "files": model_files,
        "downloaded_at": "2025-01-08"
    }
    
    with open(models_dir / "index.json", 'w') as f:
        json.dump(index_data, f, indent=2)
    
    return 0

if __name__ == "__main__":
    exit(main())